acceptance/tests/direct_puppet/supports_utf8.rb

```diff
@@ -5,14 +5,20 @@
   app_type = File.basename(__FILE__, '.*')
   tmp_environment   = mk_tmp_environment(app_type)
 
-  filenames         = {'windows' => "C:/cygwin64/tmp/#{tmp_environment}_file"}
-  filenames.default = "/tmp/#{tmp_environment}_file"
+  tmp_file = {}
+  agents.each do |agent|
+    osfamily = agent.platform =~ /windows/ ? 'windows' : 'other'
+    # just create one tmpfile per windows and other
+    tmp_file[osfamily] = agent.tmpfile(tmp_environment) unless tmp_file[osfamily]
+  end
   teardown do
     step 'remove the tmp environment symlink' do
       on master, "rm -rf #{File.join(environmentpath, tmp_environment)}"
     end
     step 'clean out produced resources' do
-      on(agents, "rm #{filenames['windows']}; rm #{filenames.default}", :accept_all_exit_codes => true)
+      agents.each_with_index do |agent, index|
+        on(agents, "rm #{tmp_file['windows']}; rm #{tmp_file['other']}", :accept_all_exit_codes => true)
```

Comment:
Keeping track of the temporary files by agent is not so bad.  Setup:

  tmp_file = {}
  agents.each do |agent|
    tmp_file[agent.hostname] = agent.tmpfile(tmp_environment)
  end

  teardown do
    step 'remove the tmp environment symlink' do
      on master, "rm -rf #{File.join(environmentpath, tmp_environment)}"
    end
    step 'clean out produced resources' do
      agents.each do |agent|
        on(agent, "rm #{tmp_file[agent.hostname]}", :accept_all_exit_codes => true)
      end
    end
  end


Here is a trick that makes the handling of the manifest easy too.  The conditional expression for the content of the temporary file can be expressed by interpolating the tmp_file hash, indexed by agent.hostname.

    manifest = <<MANIFEST
file { '#{environmentpath}/#{tmp_environment}/manifests/site.pp':
  ensure => file,
  content => '
\$test_path = \$::fqdn ? #{tmp_file}
file { \$test_path:
  content => @(UTF8)
    #{file_contents}
    | UTF8
}
  ',
}
MANIFEST


Then later, instead of indexing with ostype (windows/other), just use the agent.hostname and fact fqdn:

  step 'run agent(s)' do
    with_puppet_running_on(master, {}) do
      agents.each do |agent|
        fqdn = agent.hostname

        ...

        on(agent, "cat #{tmp_file[fqdn]}").stdout do |result|
          assert_equal(file_contents, result, 'file contents did not match accepted')


See https://gist.github.com/seangriff/de0ca02ac448bd3e740dafebde6f3c49 for complete file.